### PR TITLE
Further tests for String.Concat and String.Join

### PR DESCRIPTION
**Pull request by jamesqo**
This increases the test coverage for string.Concat and string.Join by calling `Join<T>` and `Concat<T>` with a type parameter of string, as well as adding tests for a couple of other edge cases..

cc @hughbe @stephentoub

### CODE REVIEW

File: src/System.Runtime/tests/System/StringTests.cs [hunk 1]
```diff
@@ -1388,12 +1393,19 @@ public static void Join_StringArray(string seperator, string[] values, int start
...
 
                 var iEnumerableStringNotOptimized = new Queue<string>(values);
                 Assert.Equal(expected, string.Join(seperator, iEnumerableStringNotOptimized));
+                Assert.Equal(expected, string.Join<string>(seperator, iEnumerableStringNotOptimized));
 
                 var iEnumerableObject = new List<object>(values);
                 Assert.Equal(expected, string.Join(seperator, iEnumerableObject));
+
+                // We're taking advantage of covariant arrays (which is bad) here,
+                // but this should not be a problem since Join shouldn't write to the array
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.cs]: This comment isn't necessary.

**Comment by stephentoub** [comment]
One nit, otherwise LGTM. Thanks.

**Comment by jamesqo** [comment]
I ran into a few quirks including the `Join(string, object[])` bug where it will return null if the first item is null, and another one where `Concat(object)` returns null if the `ToString` is null. Have updated the tests to account for these, everything should pass now.

**Comment by stephentoub** [comment]
LGTM. Thanks for the additions.

**Pull request merged**